[#18843] Update user-facing documentation on wait-on-conflict concurrency control (#20809)
yugabyte/yugabyte-db: docs/content/preview/architecture/transactions/concurrency-control.md

</tbody>
</table>

> ### Versioning and Upgrades

Reviewer: ```suggestion
### Versioning and upgrades
```

> This feature is upgrade and downgrade safe. When turning `enable_wait_queues` on/off, or during rolling restarts across versions with the flag “on” in the higher version, if some nodes have wait-on-conflict behavior enabled and some don’t, users will experience mixed (but still correct) behavior. A mix of both fail-on-conflict and wait-on-conflict traffic will result in the following additional YSQL specific semantics -

Reviewer: "This feature" - what feature is that?

Author: updated

> 1. If a transaction using fail-on-conflict sees transactions that have written conflicting intents -

Reviewer: ```suggestion
- If a transaction using fail-on-conflict encounters transactions that have conflicting writes:
```

>     1. If there is even 1 conflicting transaction that uses wait-on-conflict, abort the requesting transaction

Reviewer: ```suggestion
    - If there is even a single conflicting transaction that uses wait-on-conflict, the transaction aborts.
```

>     2. Else, behaviour is as today

Reviewer: Which is what? Can you be specific?

> 2. If a transaction uses wait-on-conflict and sees transactions that have written conflicting intents -

Reviewer: ```suggestion
- If a transaction using wait-on-conflict encounters transactions that have conflicting writes, it waits for all conflicting transactions to end (including any using fail-on-conflict semantics).
```

>     1. Wait for all conflicting transactions to end (including any using fail-on-conflict semantics)

Reviewer: ```suggestion
```

### Fairness

> When multiple requests are waiting on the same resource in the wait queue, and that resource becomes available, we have to decide in which order those waiting requests should get access to the contentious resource. We generally use the following process:

Reviewer: ```suggestion
When multiple requests are waiting on the same resource in the wait queue, and that resource becomes available, YugabyteDB generally uses the following process to decide in which order those waiting requests should get access to the contentious resource:
```

1. Sort all waiting requests based on the _transaction start time_, with requests from the oldest transactions first
2. Resume requests in order:

>     1. Re-run conflict resolution and acquire locks on the requested resource

Reviewer: ```suggestion
    1. Re-run conflict resolution and acquire locks on the requested resource.
```

>     2. If the resource is no longer available because another waiting request acquired it, re-enter the wait queue

Reviewer: ```suggestion
    2. If the resource is no longer available because another waiting request acquired it, re-enter the wait queue.
```

> We have two mechanisms to detect that a resource has become available:

Reviewer: ```suggestion
YugabyteDB has two mechanisms to detect that a resource has become available:

```

</tbody>
</table>

### Versioning and upgrades

When turning `enable_wait_queues` on or off, or during a rolling restart, where during an update the flag could be on on nodes with a more recent version, if some nodes have wait-on-conflict behavior enabled and some don’t, you will experience mixed (but still correct) behavior.

A mix of both fail-on-conflict and wait-on-conflict traffic results in the following additional YSQL-specific semantics:

- If a transaction using fail-on-conflict encounters transactions that have conflicting writes -
    - If there is even a single conflicting transaction that uses wait-on-conflict, the transaction aborts.

>     - Else, YugabyteDB will abort either the current request or encountered conflicting transactions based on transaction priority

Reviewer: ```suggestion
    - Otherwise, YugabyteDB uses the regular [fail-on-conflict resolution](#fail-on-conflict), which is to abort the lower priority transaction.
```
